Re-implementation of Evaluation.

dr-robertk opened the pull request:
This PR re-implements the Evaluation class to store all internal data in one array, DUNE::FieldVector is used here. This improves the code performance for flow_ebos by about 10%. However, the resulting number of linearizations, linear iterations and newton iterations differ, so this should be handled with care.

atgeirr commented:
Good idea, surprised to see that it is 10%!

dr-robertk commented:
I thinks it's that and the removal of high level functions like std::fill and std::copy which in that case are much better implemented with a simple for loop.

andlaus commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -168,15 +186,15 @@ class Evaluation
-        for (unsigned varIdx = 0; varIdx < size; ++varIdx) {
-            const ValueType& uPrime = this->derivatives_[varIdx];
-            const ValueType& vPrime = other.derivatives_[varIdx];
+        ValueType& u = data_[ valuepos_ ];
```
`const ValueType&` is sufficient here? bear in mind that copying value type objects can be expensive in the nested case. (for the usual case, i.e. primitive floating point values there should be no difference.)

dr-robertk replied:
u is used a bit below. So either const ValueType& u and then data_[ valuepos_ ] below or the way I have put it. I'm usually in favour of keeping the places where valuepos_ is used to a minimum.

andlaus commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -168,15 +186,15 @@ class Evaluation
-            const ValueType& uPrime = this->derivatives_[varIdx];
-            const ValueType& vPrime = other.derivatives_[varIdx];
+        ValueType& u = data_[ valuepos_ ];
+        const ValueType v = other.value();
```
const reference?

andlaus commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -196,15 +212,15 @@ class Evaluation
-        for (unsigned varIdx = 0; varIdx < size; ++varIdx) {
-            const ValueType& uPrime = this->derivatives_[varIdx];
-            const ValueType& vPrime = other.derivatives_[varIdx];
+        ValueType& u = data_[ valuepos_ ];
```
const

andlaus commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -196,15 +212,15 @@ class Evaluation
-            const ValueType& uPrime = this->derivatives_[varIdx];
-            const ValueType& vPrime = other.derivatives_[varIdx];
+        ValueType& u = data_[ valuepos_ ];
+        const ValueType v = other.value();
```
reference

andlaus commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -196,15 +212,15 @@ class Evaluation
+            data_[varIdx] = (v*uPrime - u*vPrime)/(v*v);
         }
-        this->value_ /= v;
+        u /= v;
```
ah: that's what prevents `u` from being `const`?

andlaus commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -295,30 +308,27 @@ class Evaluation
-
-        for (unsigned varIdx = 0; varIdx < size; ++varIdx)
-            if (this->derivatives_[varIdx] != other.derivatives_[varIdx])
+        for (int varIdx = 0; varIdx <= size; ++varIdx)
```
this is pretty non-obvious since `size` is the number of derivatives. better use `data_.size()` or maybe even `return data_ == other.data_;` works?

andlaus commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -329,53 +339,52 @@ class Evaluation
     template <class RhsValueType>
     bool operator>(RhsValueType other) const
-    { return this->value_ > other; }
+    { return data_[ valuepos_ ] > other; }
```
mabe it is better to use `value()` in such situations...

dr-robertk replied:
Bu that would be a copy of the value, right? I can check.

andlaus commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -51,19 +52,28 @@ template <class ValueT, int numVars>
 
-    enum { size = numVars };
+    //! number of derivatives
+    static constexpr int size = numVars;
```
hm. possibly the name `size` is a bit confusing. My fault, but better names are appreciated.

atgeirr replied:
I agree, what about simply exposing it as numVars (possibly renaming the template argument numVarsArg or something) or numDerivs?

andlaus commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -255,9 +268,9 @@ class Evaluation
-        result.value_ = -this->value_;
-        for (unsigned varIdx = 0; varIdx < size; ++varIdx)
-            result.derivatives_[varIdx] = - this->derivatives_[varIdx];
+        // set value and derivatives to negative
```
since these are `Dune::FieldVector`s, is it possible to use `data_ = - other.data_`. (I know, the machine code should be identical, but it would be slightly less clutter here.)

andlaus replied:
*arg*, both (and in particular the missing `operator==()`) sound awfully like a gap in `Dune::FieldVector`. anyway, thanks for the info.

atgeirr replied:
I added a free (non-class) operator- for FieldVector somewhere long ago (perhaps for CpGrid, don't remember), but that trick would not help much here. It is very annoying. I guess it is to avoid exposing temporary-creating operations, if I remember correctly?

However it prompts me to ask another thing: why not use std::array instead of Dune::FieldVector? It's an invisible internal detail and we don't gain much in terms of ease-of-writing operations for using FieldVector, it seems to me.

andlaus replied:
> I guess it is to avoid exposing temporary-creating operations, if I remember correctly?

yeah, that's true for `operator-()`, but not for `operator==()`. anyway, I don't think this is important enough to invest (the probably significant) effort to get it into Dune. In particular since we need to work around this issue anyway for several years to come.

dr-robertk commented on opm/material/densead/Evaluation.hpp:
```diff
@@ -73,65 +83,76 @@ class Evaluation
+    // derivatives being zero.
     template <class RhsValueType>
-    static Evaluation createVariable(const RhsValueType& value, unsigned varPos)
+    Evaluation(const RhsValueType& c, const int varPos)
```
What do you mean? Remove createVariable?

andlaus replied:
the additional constructor. it's new, right? (createVariable() stays around AFAICS.)

dr-robertk commented:
I re-implemented the re-implementation using std::array. It's actually slightly faster than the FieldVector version. I also addressed most of the above comments.

dr-robertk commented:
Has anybody looked at this?

andlaus commented:
> Has anybody looked at this?

yep. I'm at the final testing. expect it to be merged soon :)

atgeirr commented:
Before you merge: it was reported to change the results, and I think that should be investigated first, as @dr-robertk suggested in his initial comment. This PR is a performance optimization but is not a requirement for other work to the best of my knowledge so we need not rush this. If only flow_ebos is affected I am inclined to say fine (since it is already producing solutions that are different from flow's) but it was not clear to me if it also affected flow (which uses this class for its property evaluations).

andlaus commented:
> Before you merge: it was reported to change the results, and I think that should be investigated first, as @dr-robertk suggested in his initial comment.

I looked at this for yesterday's version: the results where the same; convergence was slightly different but the differences were not dramatic and there was no clear picture because for some decks it converged slightly better, for some slightly worse. to me it looked a lot like numerical noise. I'll test it with plain-flow before merge, though.

andlaus commented:
ok, the normal `flow` simulator produces these results for Norne on my machine:

with this PR:

```
================    End of simulation     ===============

Total time taken (seconds):  1322.9
Solver time (seconds):       1316.25
Overall Linearizations:      1952
Overall Newton Iterations:   1597
Overall Linear Iterations:   25156
```

without this PR:

```
================    End of simulation     ===============

Total time taken (seconds):  1329.96
Solver time (seconds):       1323.25
Overall Linearizations:      1976
Overall Newton Iterations:   1618
Overall Linear Iterations:   25072
```

(`flow_ebos` is at about 750 seconds.) My interpretation of these results is that convergence behaviour is slightly *different*, but overall its *not worse*.  (what can't be inferred from these summaries: the number of time step failures is about the same but they sometimes happen for different time steps.) So if there are no further objections, I'd like to merge this PR tomorrow.

atgeirr commented:
> So if there are no further objections, I'd like to merge this PR tomorrow.

Seeing as there are significant (but beneficial we hope) changes to flow results, I'd like for this to also be tested on model 2, including in parallel, before we merge this. There are also other changes already in flight that will change results, in PR OPM/opm-simulators#880. I think therefore @totto82 should give his opinion before we merge (any of these will also require updating the regression test references in opm-data).

Finally, I am a bit surprised that this can have such an effect on the results, and I'll remain nervous about that until someone can explain it. A double is a double, or perhaps not?

andlaus commented:
> I'd like for this to also be tested on model 2, including in parallel, before we merge this.

can you do this? I've never run `flow` in parallel. (since this PR does not affect the parallelization, I'd be *really* surprised if it changed things more for the parallel stuff than for the sequential runs.)

andlaus commented:
> Finally, I am a bit surprised that this can have such an effect on the results, and I'll remain nervous about that until someone can explain it. A double is a double, or perhaps not?

yeah, but in machine code, a multiplication is not always the same as a multiplication. e.g. the AVX multiplication may round the last bit differently than the x87 variant which may cascade into the differences we see.

jokva commented:
We attempted a change in the grid cell volume calculations in ert, and the difference between the optimisations before/after were on the scale of 1e-15. Still enough for flow to spit out ~1e-6 difference from its own reference solution, breaking the integration tests.

Safe to say, flow is *very* sensitive to these changes.

andlaus commented:
> Safe to say, flow is *very* sensitive to these changes.

yeah, for this reason I agree that we should not do them on release branches (or shortly before a release), but I think we should be able to do them at some point. (possibly with adapt the unit tests after looking at them manually and convinving ourselfs that there are no qualitative differences.) BTW: maybe it would be a good idea to add a CPU with a significantly different architecture (ARM?) to the testfarm?

atgeirr commented:
> can you do this? I've never run flow in parallel.

I can. Might be a little time though before I'm done (days not weeks).

atgeirr commented:
> we should be able to do them at some point.

Sure, I just want to be sure that the solution paths the simulators end up taking for our important test cases all end well before merging. What I am afraid of is a situation where for example model 2 runs suddenly cannot get past a point in the schedule, due to some bug that might be latent but only triggered by the (chaotic) minor changes.

andlaus commented:
>  I can. Might be a little time though before I'm done (days not weeks).

great, that sounds like a plan!

totto82 commented:
I have tested this PR with flow on Norne and Model2 and I get exactly the same results for both models.  
My numbers are slightly different than what @andlaus reported.   

```
Total time taken (seconds):  1304.07
Solver time (seconds):       1296.1
Overall Linearizations:      2005
Overall Newton Iterations:   1639
Overall Linear Iterations:   25440
```
Is it possible to ask Travis to check this PR from opm-simulator.

akva2 commented:
travis, no. jenkins, probably. if you explain what 'this PR' means.

totto82 commented:
@akva2 I want to test flow on Norne with PR #175 (from opm-material) and all other repos master.  The only solution I could think of is to make a trivial PR in opm-simulator and then test it with the downstream module opm-material#175. Is there something more elegant?

akva2 commented:
well you can always run it local on your machine. in top of opm-simulators at the state you want to use do

```bash
WORKSPACE=`pwd` BUILDTHREADS=<something suitable> ghprbCommentBody="jenkins build this opm-material=175 please" jenkins/build.sh
configuration=serial WORKSPACE=`pwd` jenkins/run-norne.sh
```

i'll put making the run triggers available from other repos on my TODO.

andlaus commented:
> Is there something more elegant?

maybe this:

jenkins run Norne opm-material=175 please

@akva2: as far as I know there is no trigger for running Norne with flow_ebos yet?

andlaus commented:
jenkins run Norne opm-material=175 please

akva2 commented:
when the daddy of the jenkins support says something is not supported, he probably know what he is on about ;) run triggers are only available from opm-simulators and opm-data.

and no, there is no such trigger for flow_ebos. if you want to add it, you have to replicate jenkins/run-norne.sh in opm-simulators plus i have to do some (trivial) scripting on the jenkins side.

atgeirr commented:
After discussions and testing we have a consensus that this can be merged.

andlaus commented:
ok. just like @totto82 I also get exactly the same convergence behaviour for Norne with and without this PR  when using `-O2` instead of `-Ofast -ffast-math -fno-finite-math-only` as the compiler flags. merging, thanks to all involved.

(mental note to myself: `-ffast-math` should be called `-fnondeterministic-math`.)
